dlmalloc_nonreuse.c

```diff
@@ -1195,6 +1230,59 @@ static int has_segment_link(mstate m, msegmentptr ss) {
 #define USAGE_ERROR_ACTION(m,p) ABORT
 #endif /* USAGE_ERROR_ACTION */
 
+/* --------------------------- CHERI support ----------------------------- */
+
+/*
+ * Bound a memory allocation and remove unneeded permissions.
+ */
+#ifndef __CHERI_PURE_CAPABILITY__
+#define	bound_ptr(mem, bytes)	(mem)
+#else
+static inline void *bound_ptr(void *mem, size_t bytes)
+{
+	void* ptr;
+
+	ptr = __builtin_cheri_perms_and(
+	    __builtin_cheri_bounds_set(mem, bytes),
+	    CHERI_PERMS_USERSPACE_DATA & ~CHERI_PERM_CHERIABI_VMMAP);
+	mem2chunk(mem)->pad = ptr;
+	return ptr;
+}
+#endif
+
+/*
+ * Given a memory allocation, return a pointer to it bounded to the
+ * segment it was allocated from.
+ */
+#ifndef __CHERI_PURE_CAPABILITY__
+#define	unbound_ptr(m, spp, mem)	(mem)
+#else
+#if FOOTERS
+#error We need gm when calling in free() and realloc() so can't use FOOTERS.
+#endif
+static inline void *unbound_ptr(mstate m, msegmentptr *spp, void *mem)
+{
+	msegmentptr sp;
+	void* ptr;
+	/*
+	 * Make sure we're inbounds.  Otherwise our address might match
+	 * the wrong segment.
+	 */
```

Comment:
You can elide the bounds test, but not the taggedness test, if you use the base, since that can't ever go out of bounds of the original allocation.

Reply:
I ended up deciding to check that offset == 0 since anything else is a bug anyway.  Using base in segment_holding also probably makes sense, but this lets us trivially detect another type of bug.